### ISSUE TITLE
Google analytics 5.0.0 module, crash the product page

### ISSUE THREAD

**Issue by andreicos1982** [PrestaShop/PrestaShop#35336]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

Hello. After i upgrade the google analitycs module from v 4.2.2 to v 5.0.0, the product page give error 500. I downgrade to v4.2.2 and all is ok. The prestashop version is 1.7.8.11 with php 7.4 

### Steps to reproduce

Upgrade the module from 4.2.2 to 5.0.0
Check product page is working.

### PrestaShop version(s) where the bug happened

7

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

Google analitycs v5.0.0

**Comment by oubihis** [comment]
Hey,

Have you tried activating the "[Debug mode](https://help-center.prestashop.com/en/articles/9242063447698-debug-mode)" in PrestaShop to get more details about the error? 
If not, please do so and provide the error message you receive.

**Comment by Hlavtox** [comment]
@andreicos1982 Can you please enable debug mode and tell us what the error is?

**Comment by andreicos1982** [comment]
I cant close the shop now and update the module again, but, i found the error in error log.
Stack trace:
#0 /home/r117767sani/public_html/modules/ps_googleanalytics/classes/Hook/HookDisplayFooterProduct.php(62): PrestaShop\Module\Ps_Googleanalytics\Wrapper\ProductWrapper->prepareItemFromProduct(Object(Product))
#1 /home/r117767sani/public_html/modules/ps_googleanalytics/ps_googleanalytics.php(120): PrestaShop\Module\Ps_Googleanalytics\Hooks\HookDisplayFooterProduct->run()
#2 /home/r117767sani/public_html/classes/Hook.php(1007): Ps_Googleanalytics->hookDisplayFooterProduct(Array)
#3 /home/r117767sani/public_html/classes/Hook.php(431): HookCore::coreCallHook(Object(Ps_Googleanalytics), 'hookDisplayFoot...', Array)
#4 /home/r117767sani/public_html/classes/Hook.php(944): HookCore::callHookOn(Object(Ps_Googleanalytics), 'displayFooterPr...', Array)
#5 /home/r117767sani/public_html/config/smarty.config.inc.php(167): HookCore::exec('displayFooterPr... in /home/r117767sani/public_html/modules/ps_googleanalytics/classes/Wrapper/ProductWrapper.php on line 136
[12-Feb-2024 22:02:09 Europe/Bucharest] PHP Fatal error:  Uncaught Error: Cannot use object of type Product as array in /home/r117767sani/public_html/modules/ps_googleanalytics/classes/Wrapper/ProductWrapper.php:136

**Comment by Hlavtox** [comment]
@andreicos1982 Something is messing up with the $product variable in smarty on your shop. The $product should be a lazy array, not Product object.

https://github.com/PrestaShop/ps_googleanalytics/blob/dev/classes/Hook/HookDisplayFooterProduct.php
https://github.com/PrestaShop/ps_googleanalytics/blob/dev/classes/Wrapper/ProductWrapper.php

@kpodemski 
Should I arm this module against these changes?
Also, maybe it would be better to use the product array from the hook (`{hook h='displayFooterProduct' product=$product category=$category}`), instead of the context.

**Comment by oubihis** [comment]
@andreicos1982  i think the class of ProductWrapper in ps_googleanalytics version 5.0.0 is designed to work with lazy arrays representing product data, rather than directly with Product objects or associative arrays as in version 4.2.2.

There are several ways to solve this problem without updating my theme.

 **Overridden ps_googleanalytics module**
 
You'll need to modify the ProductWrapper.php file to handle the input data correctly.
In version 5.0.0 of the ps_googleanalytics module, the ProductWrapper class expects lazy arrays representing product data, not Product objects. Therefore, you need to ensure that the input to the prepareItemFromProduct method is in the correct format.

in path : public_html/override/modules/ps_googleanalytics/classes/Wrapper/ProductWrapper.php

public function prepareItemFromProduct($product, $useProvidedQuantity = false)
{
    // Check if $product is a ProductLazyArray
    if ($product instanceof ProductLazyArray) {
        // Convert ProductLazyArray to array
        $product = $product->toArray();
    }
    
    
   // The rest of the code of ProductWrapper

_P.S : I hope I'm close to the solution because I don't have any project P.S to test_

**Comment by Hlavtox** [comment]
@oubihis @andreicos1982 Well, I could probably make the `ProductWrapper` accept the `Product` object, but I wonder how the `Product` object got there. 🤔 Because for the whole lifetime of 1.7 it's a (lazy) array.

**Comment by andreicos1982** [comment]
@oubihis I attach the product.tpl cone from my site. Maybe has something to do with som code in product.tpl?
{**
 * Copyright since 2007 PrestaShop SA and Contributors
 * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Academic Free License 3.0 (AFL-3.0)
 * that is bundled with this package in the file LICENSE.md.
 * It is also available through the world-wide-web at this URL:
 * https://opensource.org/licenses/AFL-3.0
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@prestashop.com so we can send you a copy immediately.
 *
 * DISCLAIMER
 *
 * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
 * versions in the future. If you wish to customize PrestaShop for your
 * needs please refer to https://devdocs.prestashop.com/ for more information.
 *
 * @author    PrestaShop SA and Contributors <contact@prestashop.com>
 * @copyright Since 2007 PrestaShop SA and Contributors
 * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
 *}
{extends file=$layout}
{block name='head_seo' prepend}
<link rel="canonical" href="{$product.canonical_url}">
{/block}
{block name='head' append}
<meta property="og:type" content="product">
<meta property="og:url" content="{$urls.current_url}">
<meta property="og:title" content="{$page.meta.title}">
<meta property="og:site_name" content="{$shop.name}">
<meta property="og:description" content="{$page.meta.description}">
<meta property="og:image" content="{$product.cover.large.url}">
<meta property="product:pretax_price:amount" content="{$product.price_tax_exc}">
<meta property="product:pretax_price:currency" content="{$currency.iso_code}">
<meta property="product:price:amount" content="{$product.price_amount}">
<meta property="product:price:currency" content="{$currency.iso_code}">
{if isset($product.weight) && ($product.weight != 0)}
<meta property="product:weight:value" content="{$product.weight}">
<meta property="product:weight:units" content="{$product.weight_unit}">
{/if}
{/block}
{block name='content'}
<section id="main" itemscope itemtype="https://schema.org/Product">
  <meta itemprop="url" content="{$product.url}">
  <div class="row product-detail">
    <div class="product-top-detail section-box clearfix">
      <div class="col-md-6 clearfix product-left-detail">
      {block name='page_content_container'}
        <section class="page-content clearfix" id="content">
          {block name='page_content'}
            {block name='product_flags'}
              <ul class="product-flags">
                {foreach from=$product.flags item=flag}
                  <li class="product-flag {$flag.type}">{$flag.label}</li>
                {/foreach}
              </ul>
            {/block}

            {block name='product_cover_thumbnails'}
              {include file='catalog/_partials/product-cover-thumbnails.tpl'}
            {/block}

            <div class="scroll-box-arrows">
              <i class="material-icons left">&#xE314;</i>
              <i class="material-icons right">&#xE315;</i>
            </div>
          {/block}
        </section>
      {/block}
    </div>

    <div class="col-md-6 clearfix">
      <div class="pro-right-info">
      {block name='page_header_container'}
        {block name='page_header'}
          <h1 class="h1 prdouct-title" itemprop="name">{block name='page_title'}{$product.name}{/block}</h1>
        {/block}
      {/block}

      {hook h='displayReviewsProduct'}

      {block name='product_prices'}
      {include file='catalog/_partials/product-prices.tpl'}
      {/block}

      <div class="product-information">
        {block name='product_description_short'}
          <div id="product-description-short-{$product.id}" class="product-description" itemprop="description">{$product.description_short nofilter}</div>
        {/block}


        {if $product.is_customizable && count($product.customizations.fields)}
          {block name='product_customization'}
            {include file="catalog/_partials/product-customization.tpl" customizations=$product.customizations}
          {/block}
        {/if}

        <div class="product-actions">
          {block name='product_buy'}
            <form action="{$urls.pages.cart}" method="post" id="add-to-cart-or-refresh">
              <input type="hidden" name="token" value="{$static_token}">
              <input type="hidden" name="id_product" value="{$product.id}" id="product_page_product_id">
              <input type="hidden" name="id_customization" value="{$product.id_customization}" id="product_customization_id">

              {block name='product_variants'}
              {include file='catalog/_partials/product-variants.tpl'}
              {/block}

              {block name='product_pack'}
                {if $packItems}
                  <section class="product-pack">
                    <p class="h4">{l s='This pack contains' d='Shop.Theme.Catalog'}</p>
                    {foreach from=$packItems item="product_pack"}
                      {block name='product_miniature'}
                        {include file='catalog/_partials/miniatures/pack-product.tpl' product=$product_pack}
                      {/block}
                    {/foreach}
                  </section>
                {/if}
              {/block}


              {block name='product_discounts'}
              {include file='catalog/_partials/product-discounts.tpl'}
              {/block}

              {block name='product_add_to_cart'}
              {include file='catalog/_partials/product-add-to-cart.tpl'}
              {/block}
              
              {if isset($product.reference_to_display) && $product.reference_to_display neq ''}
      <div class="product-reference">
        <label class="label">{l s='Reference' d='Shop.Theme.Catalog'} </label>
        <span itemprop="sku">{$product.reference_to_display}</span>
      </div>
    {/if}

              {block name='product_additional_info'}
              {include file='catalog/_partials/product-additional-info.tpl'}
              {/block}

              {* Input to refresh product HTML removed, block kept for compatibility with themes *}
              {block name='product_refresh'}{/block}
            </form>
          {/block}
          
        </div>

        {block name='hook_display_reassurance'}
          {hook h='displayReassurance'}
        {/block}
      </div>
      </div>
    </div>
    </div>

    {block name='product_tabs'}
      <div class="col-md-12">
        <div class="tabs section-margin-top">
          <ul class="nav nav-tabs" role="tablist">
            {if $product.description}
              <li class="nav-item">
                <a
                class="nav-link{if $product.description} active{/if}"
                data-toggle="tab"
                href="#description"
                role="tab"
                aria-controls="description"
                {if $product.description} aria-selected="true"{/if}>{l s='Description' d='Shop.Theme.Catalog'}</a>
              </li>
            {/if}
            <li class="nav-item">
              <a
              class="nav-link{if !$product.description} active{/if}"
              data-toggle="tab"
              href="#product-details"
              role="tab"
              aria-controls="product-details"
              {if !$product.description} aria-selected="true"{/if}>{l s='Product Details' d='Shop.Theme.Catalog'}</a>
            </li>

            {hook h='ProductTab'}

            {if $product.attachments}
              <li class="nav-item">
                <a
                class="nav-link"
                data-toggle="tab"
                href="#attachments"
                role="tab"
                aria-controls="attachments">{l s='Attachments' d='Shop.Theme.Catalog'}</a>
              </li>
            {/if}
            {foreach from=$product.extraContent item=extra key=extraKey}
              <li class="nav-item">
                <a
                class="nav-link"
                data-toggle="tab"
                href="#extra-{$extraKey}"
                role="tab"
                aria-controls="extra-{$extraKey}">{$extra.title}</a>
              </li>
            {/foreach}
          </ul>
          <div class="tab-content" id="tab-content">
            <div class="tab-pane fade in{if $product.description} active{/if}" id="description" role="tabpanel">
              {block name='product_description'}
              <div class="product-description">{$product.description nofilter}</div>
              {/block}
            </div>

            {block name='product_details'}
            {include file='catalog/_partials/product-details.tpl'}
            {/block}

            <div class="tab-pane fade in" id="tab-review" role="tabpanel">
              {hook h='productTabContent'}
            </div>

            {block name='product_attachments'}
              {if $product.attachments}
                <div class="tab-pane fade in" id="attachments" role="tabpanel">
                  <section class="product-attachments">
                    <p class="h5 text-uppercase">{l s='Download' d='Shop.Theme.Actions'}</p>
                    {foreach from=$product.attachments item=attachment}
                      <div class="attachment">
                        <h4><a href="{url entity='attachment' params=['id_attachment' => $attachment.id_attachment]}">{$attachment.name}</a></h4>
                        <p>{$attachment.description}</p>
                        <a href="{url entity='attachment' params=['id_attachment' => $attachment.id_attachment]}">
                        {l s='Download' d='Shop.Theme.Actions'} ({$attachment.file_size_formatted})
                        </a>
                      </div>
                    {/foreach}
                  </section>
                </div>
              {/if}
            {/block}

            {foreach from=$product.extraContent item=extra key=extraKey}
              <div class="tab-pane fade in {$extra.attr.class}" id="extra-{$extraKey}" role="tabpanel" {foreach $extra.attr as $key => $val} {$key}="{$val}"{/foreach}>
              {$extra.content nofilter}
              </div>
            {/foreach}
          </div>
        </div>
      </div>
    {/block}
  </div>
  {block name='product_accessories'}
    {if $accessories}
      <section class="product-accessories section-margin-top clearfix">
        <div class="section-box clearfix">
          <h2 class="products-section-title"><span>{l s='Related Products' d='Shop.Theme.Catalog'}</span></h2>
          <div class="row">
            <div class="products product-accessories-carousel" data-items="5">
            {foreach from=$accessories item="product_accessory"}
              {block name='product_miniature'}
                {include file='catalog/_partials/miniatures/product.tpl' product=$product_accessory}
              {/block}
            {/foreach}
          </div>
          </div>
        </div>
      </section>
    {/if}
  {/block}

  {block name='product_footer'}
    {hook h='displayFooterProduct' product=$product category=$category}
  {/block}

  {block name='product_images_modal'}
    {include file='catalog/_partials/product-images-modal.tpl'}
  {/block}

  {block name='page_footer_container'}
    <footer class="page-footer">
      {block name='page_footer'}
        
      {/block}
    </footer>
  {/block}
</section>
{/block}

**Comment by andreicos1982** [comment]
Ok. I found out the problem. Is because of product comment module. If i disable the module all is working.

**Comment by florine2623** [comment]
I'm using PS 8.1.x and develop, with module ps_googleanalytics v5.0.0.

I do not have any issue on the product page on BO nor FO.


https://github.com/PrestaShop/PrestaShop/assets/16019289/ddbbcc3c-08c6-43a7-bc86-3254fb8062dd

I'm closing the issue as you found a solution.

**Comment by edgarcelso** [comment]
> Hello. After i upgrade the google analitycs module from v 4.2.2 to v 5.0.0, the product page give error 500. I downgrade to v4.2.2 and all is ok. The prestashop version is 1.7.8.11 with php 7.4

i have the same problem, prestashop version 1.7.8.8 with php 7.4.
i don't know how to downgrade the module, so i uninstalled and the product page is fine


<img width="1008" alt="Captura de ecrã 2024-02-25 234427" src="https://github.com/PrestaShop/PrestaShop/assets/6127057/3a631110-a72b-4920-a2d0-062feaee8c9d">
<img width="1042" alt="Captura de ecrã 2024-02-25 234454" src="https://github.com/PrestaShop/PrestaShop/assets/6127057/ef5212db-070e-4d1e-b2b6-be458fc905c8">

**Comment by rycardo74** [comment]
Google Analytics v5.0.2 - by PrestaShop.

AH01071: Got error 'PHP message: PHP Fatal error: Uncaught Error: Cannot use object of type Product as array

**Comment by Hlavtox** [comment]
@rycardo74 Something is redefining the $product object in template. 👍

**Comment by kpodemski** [comment]
@Hlavtox maybe we could change the module and rely on `$this->context->controller->getProduct()`? it's not the first person reporting this issue, better to anticipate it

**Comment by Hlavtox** [comment]
@kpodemski The module is taking a normal variable from displayFooterProduct, the Product object is not used in FO at all, it should not be there. There is some module that is overriding it...

Btw, `$this->context->controller->getProduct()` will give you non-presented product object, we are using a final lazy array in the module, to save performance. The module now has virtually zero overhead. 👍

**Comment by rycardo74** [comment]
where can I change this line?

**Comment by kpodemski** [comment]
@Hlavtox 

Daniel, please, this problem with overridden product variable is known for ages, at least we could check if the input is what we really expect, and if not do something to mitigate the issue

**Comment by Kandriws** [comment]
I encountered an issue with the Google Analytics module in PrestaShop. Here’s how I resolved it:

**Steps to Fix:**

1. Open the file: `public_html/modules/ps_googleanalytics/classes/Wrapper/ProductWrapper.php`.
2. Locate the `prepareItemFromProduct` function in the code.
3. Add the following code within this function:

```php
if (!($product instanceof ProductListingLazyArray)) {
    $product = (array) $product;
}
```

**Explanation:**

This code snippet checks if the `$product` is an instance of `ProductListingLazyArray`. If it is not, the code converts `$product` to a native array. This conversion resolves the issue with the Google Analytics module.

With this fix, I was able to resolve the bug.
